### PR TITLE
Add tests/docs for Pinned Memory metrics

### PR DESCRIPTION
**Pull request by pskiran1**
core: https://github.com/triton-inference-server/core/pull/306

### CODE REVIEW

File: qa/L0_metrics/pinned_memory_metrics_test.py [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+        ]
+
+        self.outputs = [
+            httpclient.InferRequestedOutput("OUTPUT__0"),
+            httpclient.InferRequestedOutput("OUTPUT__1"),
+        ]
+
+        # Before loading the model
+        total_bytes_value, used_bytes_value = self._get_metrics()
+        self.assertEqual(int(total_bytes_value), 268435456)
```
Comment (rmccorm4) [inline qa/L0_metrics/pinned_memory_metrics_test.py:67]: Can you define a readable constant for this expected value?

Comment (rmccorm4) [reply qa/L0_metrics/pinned_memory_metrics_test.py:67]: Can you also try modifying the pinned memory pool size on server side, and have the test confirm the size is correct too? You can communicate the expected size over an environment variable or something for the test.

Comment (pskiran1) [reply qa/L0_metrics/pinned_memory_metrics_test.py]: Done 👍

File: qa/L0_metrics/pinned_memory_metrics_test.py [hunk 2]
```diff
@@ -0,0 +1,177 @@
+#!/bin/bash
```
Comment (rmccorm4) [inline qa/L0_metrics/pinned_memory_metrics_test.py:1]: `#!/usr/bin/python`

File: qa/L0_metrics/pinned_memory_metrics_test.py [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+import requests
+import tritonclient.http as httpclient
+from tritonclient.utils import *
+
+custom_pinned_memory_pool_size = os.environ.get("CUSTOM_PINNED_MEMORY_POOL_SIZE")
+if custom_pinned_memory_pool_size is not None:
+    TOTAL_PINNED_MEMORY_SIZE = int(custom_pinned_memory_pool_size)
+else:
+    # Triton server reserves 256 MB for pinned memory by default.
+    TOTAL_PINNED_MEMORY_SIZE = 2**28  # bytes, Equivalent to 256 MB
```
Comment (rmccorm4) [inline qa/L0_metrics/pinned_memory_metrics_test.py:44]: Can use default value behavior built-in to `.get(value, default)`
```suggestion
DEFAULT_TOTAL_PINNED_MEMORY_SIZE = 2**28 # bytes, Equivalent to 256 MB
TOTAL_PINNED_MEMORY_SIZE = int(os.environ.get("CUSTOM_PINNED_MEMORY_POOL_SIZE", DEFAULT_TOTAL_PINNED_MEMORY_SIZE))
```

File: qa/L0_metrics/pinned_memory_metrics_test.py [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+print(f"TOTAL_PINNED_MEMORY_SIZE: {TOTAL_PINNED_MEMORY_SIZE} bytes")
+
+# Pinned memory usage when server is idle (no inference)
+DEFAULT_USED_PINNED_MEMORY_SIZE = 0  # bytes
+
+
+total_bytes_pattern = re.compile(r"pool_total_bytes (\d+)")
+used_bytes_pattern = re.compile(r"pool_used_bytes (\d+)")
```
Comment (rmccorm4) [inline qa/L0_metrics/pinned_memory_metrics_test.py:52]: Can move this inside `_get_metrics()` if not used elsewhere

File: qa/L0_metrics/pinned_memory_metrics_test.py [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+                        model_name=self.model_name,
+                        inputs=self.inputs,
+                        outputs=self.outputs,
+                    )
+                )
+
+            time.sleep(1)
+
+            # Set the event to indicate that inference is completed
+            self.inference_completed.set()
```
Comment (rmccorm4) [inline qa/L0_metrics/pinned_memory_metrics_test.py:134]: Inference isn't complete in async case until all the results are finished, so I think this should be moved after all the `get_result()` have finished.

**Review by tanmayv25 — APPROVED** [review]

**Review by rmccorm4 — APPROVED** [review]

**Pull request merged**